### PR TITLE
BUG: Fix Categorical use_inf_as_na bug

### LINKED ISSUE #33594: BUG: pandas 1.0 dropna error with categorical data if pd.options.mode.use_inf_as_na = True
- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import pandas as pd
import numpy as np
from pandas.api.types import CategoricalDtype

# with categorical column and use_inf_as_na = True -> ERROR
pd.options.mode.use_inf_as_na = True
df1 = pd.DataFrame([['a1', 'good'], ['b1', 'good'], ['c1', 'good'], ['d1', 'bad']], columns=['C1', 'C2'])
df2 = pd.DataFrame([['a1', 'good'], ['b1', np.inf], ['c1', np.NaN], ['d1', 'bad']], columns=['C1', 'C2'])
categories = CategoricalDtype(categories=['good', 'bad'], ordered=True)
df1.loc[:, 'C2'] = df1['C2'].astype(categories)
df2.loc[:, 'C2'] = df2['C2'].astype(categories)
df1.dropna(axis=0)  # ERROR
df2.dropna(axis=0)  # ERROR
```

#### Problem description

With the latest version of pandas (1.0.3, installed via pip on python 3.6.8), DataFrame.dropna returns an error if a column is of type **CategoricalDtype** AND **pd.options.mode.use_inf_as_na = True**. 

Exception with traceback:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/sbucc/miniconda3/envs/tf114/lib/python3.6/site-packages/pandas/core/frame.py", line 4751, in dropna
    count = agg_obj.count(axis=agg_axis)
  File "/home/sbucc/miniconda3/envs/tf114/lib/python3.6/site-packages/pandas/core/frame.py", line 7800, in count
    result = notna(frame).sum(axis=axis)
  File "/home/sbucc/miniconda3/envs/tf114/lib/python3.6/site-packages/pandas/core/dtypes/missing.py", line 376, in notna
    res = isna(obj)
  File "/home/sbucc/miniconda3/envs/tf114/lib/python3.6/site-packages/pandas/core/dtypes/missing.py", line 126, in isna
    return _isna(obj)
  File "/home/sbucc/miniconda3/envs/tf114/lib/python3.6/site-packages/pandas/core/dtypes/missing.py", line 185, in _isna_old
    return obj._constructor(obj._data.isna(func=_isna_old))
  File "/home/sbucc/miniconda3/envs/tf114/lib/python3.6/site-packages/pandas/core/internals/managers.py", line 555, in isna
    return self.apply("apply", func=func)
  File "/home/sbucc/miniconda3/envs/tf114/lib/python3.6/site-packages/pandas/core/internals/managers.py", line 442, in apply
    applied = getattr(b, f)(**kwargs)
  File "/home/sbucc/miniconda3/envs/tf114/lib/python3.6/site-packages/pandas/core/internals/blocks.py", line 390, in apply
    result = func(self.values, **kwargs)
  File "/home/sbucc/miniconda3/envs/tf114/lib/python3.6/site-packages/pandas/core/dtypes/missing.py", line 183, in _isna_old
    return _isna_ndarraylike_old(obj)
  File "/home/sbucc/miniconda3/envs/tf114/lib/python3.6/site-packages/pandas/core/dtypes/missing.py", line 283, in _isna_ndarraylike_old
    vec = libmissing.isnaobj_old(values.ravel())
TypeError: Argument 'arr' has incorrect type (expected numpy.ndarray, got Categorical)

This doesn't happen with pandas 0.24.0 or if pd.options.mode.use_inf_as_na = False (default).

#### Expected Output

no error

#### Output of ``pd.show_versions()``

<details>

>>> pd.show_versions()

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.6.8.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.15.0-96-generic
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 1.0.3
numpy            : 1.18.2
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 46.1.3.post20200330
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : 3.2.1
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : None
pyxlsb           : None
s3fs             : None
scipy            : 1.4.1
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by dsaxton**
- [x] closes #33594
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/dtypes/missing.py [hunk 1]
```diff
@@ -231,7 +231,11 @@ def _isna_ndarraylike(obj):
 
 
 def _isna_ndarraylike_old(obj):
-    values = getattr(obj, "_values", obj)
+    if not isinstance(obj, np.ndarray):
+        values = obj.to_numpy()
+    else:
+        values = obj
```
Comment (dsaxton) [inline pandas/core/dtypes/missing.py:237]: @simonjayhawkins How would we get mypy to catch bugs like this? The problem was evidently that values (Categorical) got passed to _isna_string_dtype which expects an ndarray according to annotations

Comment (WillAyd) [reply pandas/core/dtypes/missing.py:237]: Right now anything in the np.ndarray space resolved to Any, so we don't get type checking against it. Most likely need stubs or annotations for numpy to have this catch going forward

**Review by WillAyd — COMMENTED** [review]
Looks reasonable @TomAugspurger

File: pandas/core/dtypes/missing.py (hunk 1 above)
Comment (dsaxton) [reply pandas/core/dtypes/missing.py:237]: Thanks @WillAyd. Is having np.* resolve to Any a conscious decision until more types get clarified?

Comment (WillAyd) [reply pandas/core/dtypes/missing.py:237]: This is just how mypy works when you set the `ignore_missing_imports` option (see setup.cfg); otherwise mypy would fail that there are no annotations found in the np.* space

File: pandas/core/dtypes/missing.py [hunk 2]
```diff
@@ -231,7 +231,11 @@ def _isna_ndarraylike(obj):
 
 
 def _isna_ndarraylike_old(obj):
-    values = getattr(obj, "_values", obj)
+    if not isinstance(obj, np.ndarray):
```
Comment (jreback) [inline pandas/core/dtypes/missing.py:234]: is the the same construct we use in isna_arraylike?

File: pandas/_libs/missing.pyx [hunk 3]
```diff
@@ -160,7 +161,11 @@ def isnaobj_old(arr: ndarray) -> ndarray:
     result = np.zeros(n, dtype=np.uint8)
     for i in range(n):
         val = arr[i]
-        result[i] = val is NaT or _check_none_nan_inf_neginf(val)
+        result[i] = (
```
Comment (jreback) [inline pandas/_libs/missing.pyx:164]: is this the sam construct we have in isnaobj?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/dtypes/missing.py (hunk 2 above)
Comment (dsaxton) [reply pandas/core/dtypes/missing.py:234]: In _isna_arraylike it looks like we use getattr, but then check if we have an ExtensionArray and call the na method if so

File: pandas/_libs/missing.pyx (hunk 3 above)
Comment (dsaxton) [reply pandas/_libs/missing.pyx:164]: We use checknull there, I'll try to incorporate that here

File: pandas/_libs/missing.pyx [hunk 4]
```diff
@@ -160,7 +161,7 @@ def isnaobj_old(arr: ndarray) -> ndarray:
     result = np.zeros(n, dtype=np.uint8)
     for i in range(n):
         val = arr[i]
-        result[i] = val is NaT or _check_none_nan_inf_neginf(val)
+        result[i] = checknull(val) or val == INF or val == NEGINF
```
Comment (simonjayhawkins) [inline pandas/_libs/missing.pyx:164]: Is this related to the OP in #33594?. I _think_ the NA checking for string dtype https://github.com/pandas-dev/pandas/issues/33594#issuecomment-615879369 maybe should be a separate issue/PR. We _may_ want to backport the fix relating to the change in return type of Categorical.ravel from numpy array to Categorical.

File: pandas/core/dtypes/missing.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/core/dtypes/missing.py:234]: I agree the code should use the same structure as `_isna_ndarraylike`, refactoring out the common code maybe advantageous to avoid divergence occurring again.

File: pandas/_libs/missing.pyx (hunk 4 above)
Comment (dsaxton) [reply pandas/_libs/missing.pyx:164]: It's unrelated, I found this other bug while debugging the original issue and decided to fix it here. I'll go ahead and separate it out for easier bookkeeping.

File: pandas/core/dtypes/missing.py (hunk 2 above)
Comment (dsaxton) [reply pandas/core/dtypes/missing.py:234]: How would you recommend using a similar pattern here? We can't rely on isna to catch infinity (it only works if you construct the array _after_ setting the option):

```python
[ins] In [1]: cat = pd.Categorical([1, 2, np.inf])

[ins] In [2]: pd.options.mode.use_inf_as_na = True

[ins] In [3]: cat.isna()
Out[3]: array([False, False, False])

[ins] In [4]: cat = pd.Categorical([1, 2, np.inf])

[ins] In [5]: cat.isna()
Out[5]: array([False, False,  True])
```

File: pandas/_libs/missing.pyx (hunk 4 above)
Comment (dsaxton) [reply pandas/_libs/missing.pyx]: Moved to https://github.com/pandas-dev/pandas/pull/33656

Comment (dsaxton) [reply pandas/_libs/missing.pyx]: Actually, they are kind of related in that a fix needs to be made missing.py as well. I'll wait to see how that ends up here before making the associated change in the other PR.

Comment (simonjayhawkins) [reply pandas/_libs/missing.pyx]: one is a bug in `_isna_ndarraylike_old` and the other is a bug in `isnaobj_old`. They are related but I think should be treated indepedently.

for the NA issue (bug in `isnaobj_old` once bug in `_isna_ndarraylike_old` is fixed) we may want to add more tests for other extension arrays or add a test/parameterise an existing test in the base extension array tests.

File: pandas/core/dtypes/missing.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/core/dtypes/missing.py:234]: so is there also a bug in Categorical.isna?

Comment (dsaxton) [reply pandas/core/dtypes/missing.py:234]: I suppose you could call this behavior a bug, but I'm not sure if it's a problem with Categorical.isna (it's simply checking Categorical._codes, which makes total sense and shouldn't need to change in my opinion). IMHO the issue is more with trying to have the behavior of these constructors / methods respect hidden global variables instead of being well-defined in isolation (I could even see the merit of deprecating options like this rather than needing to patch every place where NA checking might happen)

Comment (jreback) [reply pandas/core/dtypes/missing.py:234]: I would prefer to make this code substatially the same, difference only be the inf change.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/dtypes/missing.py [hunk 5]
```diff
@@ -233,9 +233,10 @@ def _isna_ndarraylike_old(obj):
     values = getattr(obj, "_values", obj)
     dtype = values.dtype
 
-    if is_string_dtype(dtype):
+    if is_extension_array_dtype(dtype):
+        result = values.isna() | (values == -np.inf) | (values == np.inf)
```
Comment (TomAugspurger) [inline pandas/core/dtypes/missing.py:237]: Is this always treating inf as NA?

Comment (dsaxton) [reply pandas/core/dtypes/missing.py:237]: Yes, I think / hope that's the intention when this gets called (e.g., at line 244 we have `result = ~np.isfinite(values)`)

**Review by jorisvandenbossche — APPROVED** [review]

File: pandas/tests/arrays/categorical/test_missing.py [hunk 6]
```diff
@@ -97,3 +98,26 @@ def test_fillna_array(self):
...
+        [
+            ([1, 2, 3], np.array([False, False, False])),
+            ([1, 2, np.nan], np.array([False, False, True])),
+            ([1, 2, np.inf], np.array([False, False, True])),
+        ],
+    )
+    def test_use_inf_as_na(self, values, expected):
+        # https://github.com/pandas-dev/pandas/issues/33594
+        with pd.option_context("mode.use_inf_as_na", True):
+            cat = Categorical(values)
```
Comment (jorisvandenbossche) [inline pandas/tests/arrays/categorical/test_missing.py:113]: Should we also test this with putting the Categorical creation *outside* of the option context?

Comment (dsaxton) [reply pandas/tests/arrays/categorical/test_missing.py:113]: It turns out that actually fails interestingly enough: https://github.com/pandas-dev/pandas/pull/33629#discussion_r410936224 (if you set the option after constructing the object it loses its effect)

Comment (jorisvandenbossche) [reply pandas/tests/arrays/categorical/test_missing.py:113]: > What's even more strange is the value displays as NaN:

I think that is somewhat "expected", given the discussion above in https://github.com/pandas-dev/pandas/pull/33629#discussion_r410936224 (not that I like that behaviour though).

But regardless of how the categorical is created (with or without the option), I find it very strange that `isna` is then failing

Comment (jorisvandenbossche) [reply pandas/tests/arrays/categorical/test_missing.py:113]: Hmm, being a bit confused. 
The Categorical.isna is failing, of course, since it just looks at -1 in the codes, and when inf is part of the categories, it is not -1 in the codes. Hence it is failing to detect it.

File: pandas/core/dtypes/missing.py (hunk 5 above)
Comment (jreback) [reply pandas/core/dtypes/missing.py:237]: can you make the similar code paths in isna_ndarraylike an here

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/dtypes/missing.py (hunk 5 above)
Comment (dsaxton) [reply pandas/core/dtypes/missing.py:237]: The functions are almost the same now, unless there are objections (can revert if so) I'm going to try parametrizing _isnda_ndarraylike over old / new and just deleting this function

File: pandas/tests/arrays/categorical/test_missing.py (hunk 6 above)
Comment (jorisvandenbossche) [reply pandas/tests/arrays/categorical/test_missing.py:113]: @dsaxton so could you additionally test it for `pd.isna` ? In constract to Categorical.isna(), pd.isna should work even if the Categorical is created before the context, I think?

Comment (dsaxton) [reply pandas/tests/arrays/categorical/test_missing.py:114]: @jorisvandenbossche Added some tests, let me know if this is roughly what you had in mind

File: pandas/tests/arrays/categorical/test_missing.py [hunk 7]
```diff
@@ -97,3 +98,59 @@ def test_fillna_array(self):
...
+    def test_use_inf_as_na_outside_context(self, values, expected):
+        # https://github.com/pandas-dev/pandas/issues/33594
+        # Using isna directly for Categorical will fail in general here
+        cat = Categorical(values)
+
+        with pd.option_context("mode.use_inf_as_na", True):
+            result = pd.isna(cat)
+            tm.assert_numpy_array_equal(result, expected)
+
+        cat = pd.Series(Categorical(values))
```
Comment (jorisvandenbossche) [inline pandas/tests/arrays/categorical/test_missing.py:144]: This is not needed outside of the context, since it only matters that the categorical is created outside (the same for dataframe)

File: pandas/tests/arrays/categorical/test_missing.py (hunk 6 above)
Comment (jorisvandenbossche) [reply pandas/tests/arrays/categorical/test_missing.py:114]: Yes, that looks good!

File: pandas/tests/arrays/categorical/test_missing.py (hunk 7 above)
Comment (dsaxton) [reply pandas/tests/arrays/categorical/test_missing.py:144]: Ok, wasn't sure if somehow the NAs might get reconstructed, will just use a single context

**Review by jorisvandenbossche — APPROVED** [review]

File: pandas/core/dtypes/missing.py [hunk 8]
```diff
@@ -207,40 +207,25 @@ def _use_inf_as_na(key):
         globals()["_isna"] = _isna_new
 
 
-def _isna_ndarraylike(obj):
+def _isna_ndarraylike(obj, old=False):
```
Comment (jreback) [inline pandas/core/dtypes/missing.py:210]: can you add a doc-string here and type old

**Review by jreback — CHANGES_REQUESTED** [review]
small change, otherwise lgtm. ping on green.

in a followon, I'd like to remove entirely the _isna_old and push just have a single parameterized one as you can just parameterize _isna_new / _old

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @dsaxton 

if you can push the followon when you can to consolidate code here would be great

**Comment by simonjayhawkins** [comment]
xref #33300 to track.
